testcase/cross-region/cross-region.go

```diff
@@ -0,0 +1,216 @@
+
+	"github.com/pingcap/tipocket/pkg/cluster"
+	"github.com/pingcap/tipocket/pkg/core"
+	"github.com/pingcap/tipocket/pkg/util/pdutil"
+	"github.com/pingcap/tipocket/util"
+)
+
+const stmtDrop = `
+drop table if exists t;
+`
+const stmtCreate = `
+create table t (c int)
+PARTITION BY RANGE (c) (
+	PARTITION p1 VALUES LESS THAN (6),
+	PARTITION p2 VALUES LESS THAN (11),
+	PARTITION p3 VALUES LESS THAN (16)
+);`
+
+const stmtPolicy = `
+alter table t alter partition %v
+add placement policy
+	constraints='["+zone=%v"]'
+	role=leader
+	replicas=1`
+
+// Config exposes the config
+type Config struct {
+	TSORequests int
+	DBName      string
+	//TODO: support configure DCLocationNum instead of fixed 6
+	DCLocationNum int
+}
+
+// ClientCreator creates crossRegionClient
+type ClientCreator struct {
+	Cfg *Config
+}
+
+// Create ...
+func (l ClientCreator) Create(node cluster.ClientNode) core.Client {
+	return &crossRegionClient{
+		Config:      l.Cfg,
+		initialized: false,
+		closed:      false,
+	}
+}
+
+type crossRegionClient struct {
+	*Config
+	pdClient     pdClient.Client
+	pdHTTPClient *pdutil.Client
+	db           *sql.DB
+	initialized  bool
+	closed       bool
+}
+
+// SetUp...
+func (c *crossRegionClient) SetUp(ctx context.Context, _ []cluster.Node, cnodes []cluster.ClientNode, idx int) error {
+	if c.initialized {
+		return nil
+	}
+	name := cnodes[idx].ClusterName
+	namespace := cnodes[idx].Namespace
+	pdAddr := buildPDSvcName(name, namespace)
+	if c.pdHTTPClient == nil {
+		c.pdHTTPClient = pdutil.NewPDClient(http.DefaultClient, "http://"+pdAddr)
+	}
+	dsn := fmt.Sprintf("root@tcp(%s)/%s", buildTiDBSvcName(name, namespace), c.DBName)
+	db, err := util.OpenDB(dsn, 20)
+	if err != nil {
+		return fmt.Errorf("[on_dup] create db client error %v", err)
+	}
+	c.db = db
+	err = c.setup()
+	if c.pdClient == nil {
+		c.pdClient, err = pdClient.NewClientWithContext(ctx, []string{pdAddr}, pdClient.SecurityOption{})
+		if err != nil {
+			return err
+		}
+	}
+	if err != nil {
+		return err
+	}
+	c.initialized = true
+	return nil
+}
+
+// TearDown...
+func (c *crossRegionClient) TearDown(ctx context.Context, _ []cluster.ClientNode, idx int) error {
+	if c.closed {
+		return nil
+	}
+	if c.db != nil {
+		c.db.Close()
+	}
+	if c.pdClient != nil {
+		c.pdClient.Close()
+	}
+	c.closed = true
+	return nil
+}
+
+// Start...
+func (c *crossRegionClient) Start(ctx context.Context, cfg interface{}, cnodes []cluster.ClientNode) error {
+	return c.testTSO(ctx)
+}
+
+func (c *crossRegionClient) setup() error {
+	err := c.setupPD()
+	if err != nil {
+		return err
+	}
+	log.Info("PD setup successfully")
+	err = c.setupStore()
+	if err != nil {
+		return err
+	}
+	log.Info("TiKV setup successfully")
+	err = c.setupDB()
+	if err != nil {
+		return err
+	}
+	log.Info("DB setup successfully")
+	return nil
+}
+
+func (c *crossRegionClient) setupPD() error {
+	members, err := c.pdHTTPClient.GetMembers()
+	if err != nil {
+		return err
+	}
+	if len(members.Members) != 6 {
```

Comment:
I suggest using `GetMembers()` rather than field name to prevent potential panic cases.

Reply:
The `members` here was mocked, not by `kvproto`.